### ISSUE TITLE
Feature Request: Notification API

### ISSUE THREAD

**Issue by jag-k** [raycast/utils#15]
Please, add a Notification API support 🙏

**Comment by SKaplanOfficial** [comment]
This is an overall good idea, imo, but it might be more worthwhile to point people toward [node-notifier](https://github.com/mikaelbr/node-notifier). Apple’s UserNotifications framework has some annoying limitations, such as not being able to (easily) customize the notification icon, which node-notifier/terminal-notifier have already overcome. A custom implementation for Raycast would probably use the same tactics in the end (i.e. using a separate app bundle for the notification system and dynamically updating the icon within that app’s bundle).

Perhaps a reasonable solution would be to build an interface on top of node-notifier so that developers can use Images and Colors from the Raycast API.

**Comment by mathieudutour** [comment]
What would be the use case for it, and how would it differ from the Toast/HUD API that we provide?

**Comment by mathieudutour** [comment]
What kind of messages? Why are the `HUD` not enough?

**Comment by SKaplanOfficial** [comment]
One use case for me would be to provide actionable updates on async/long-running tasks, for example in my Image Modification extension I'd like to provide notifications when conversion/filter application/etc. is finished, allowing users to click the notification and open the result, copy it to the clipboard, and so on. Would also be useful to provide ways to recover from failures without requiring the user's constant focus on the Raycast window.

`HUD` doesn't provide the interactivity I'd like, and `Toast` isn't viable in the situation I'm envisioning. Native notifications seem like a good solution as they don't require re-inventing the wheel, users are already familiar with how they function, and using them aligns well with Raycast's emphasis on being a native macOS app.

**Comment by mathieudutour** [comment]
The issue with notification is with regards to the lifecycle of the extensions. We would have to keep the context of the extension in memory until the notification is cleared. And that's not something we want to do.

Something we could envision is being able to click on the HUD to "transform" it (eg. open raycast and show it as a toast) to be able to add actions and more details

**Comment by SKaplanOfficial** [comment]
I see your point, though I wonder if Deeplinks with Launch Context already provide a way to avoid some of those headaches -- i.e. instead of keeping the extension loaded in memory, offload it and have notifications start a new session of the extension with data communicated via launch context, and maybe a new `notification` launch type.

That said, I'd be happy with an updated HUD as you mentioned. I mostly just want a way to intuitively surface available options to users.

**Comment by KevinBatdorf** [comment]
I use notifications in a background command to notify the user when something happens (in this case when their code is deploying). Notifications are expected behavior for something like this and arent as distracting in my opinion as the HUD notice appearing.

Currently I fire a notification like this, but something native with a custom icon and actions would be nice (or even without the actions): https://github.com/KevinBatdorf/laravel-forge-raycast/blob/44fa7f97be4c69736578ef1ce9b075eda3b36755/src/check-deploy-status.tsx#L36

But yeah, I have no idea the technical challenges in supporting that at the OS level. Just giving feedback.

**Comment by mathieudutour** [comment]
Ideally, we don't want the dev to worry about what kind of mechanism they use to give feedback to the user. It should be a implementation detail and Raycast should take care of showing what's the most appropriate. 

> Notifications are expected behavior for something like this and arent as distracting in my opinion as the HUD notice appearing.

That's exactly what we don't want to happen, 1/3 of the extensions using toast, 1/3 using HUD, 1.3 using notifications - without the user knowing what to expect.

What we will probably end up doing is deprecating `showHUD` in favour of a single `showToast` which will handle the presentation mode under the hood

**Comment by mathieudutour** [comment]
> Raycast provides two different experiences though, one with user interaction, and another in the background.

Yeah but from a dev point of view, it shouldn't matter - Raycast should know better what to do to be consistent
